Draft extension for SPZ compression support for Gaussian splats (#81)
CesiumGS/glTF: extensions/2.0/Khronos/KHR_spz_compression/README.md

> # KHR\_spz\_compression

Reviewer: We need a different name for this. My current favorite is `KHR_gaussian_splatting_spz_compression` but it might be too long. The shorter `KHR_gsplat_spz_compression` might be better.

Reviewer: We met with @lilleyse and decided on a name: `KHR_spz_gaussian_splats_compression`

## Contributors

- Jason Sobotka, Cesium
- Adam Morris, Cesium
- Sean Lilley, Cesium

## Status

Draft

## Dependencies

Written against the glTF 2.0 spec.

## Table of Contents

- [Overview](#overview)

> - [Adding Gaussian Splats](#adding-spz-compressed-data-to-primitives)

Reviewer: Should we be saying `Gaussian Splatting` or `Gaussian Splats`?

Reviewer: We decided on `Gaussian Splats`.

  - [Extension Properties](#extension-properties)
  - [Accessors](#accessors)
- [Conformance](#conformance)
- [Implementation](#implementation)
- [Schema](#schema)
- [Known Implementations](#known-implementations)
- [Resources](#resources)

## Overview

> 

Reviewer: At the start of the overview, we need to add a lot of context around Gaussian Splatting. We can probably reuse most if not all of the overview in the old document. Remember that this specification is not just adding SPZ support, but also enabling Gaussian Splatting within glTF.

> SPZ is a compression format from Niantic Labs designed for Gaussian splats. Currently, it is open sourced under the MIT license. The SPZ format is primarily used in the Niantic Labs Scaniverse app. It was purpose built for Guassian splats and offers a balance of high compression with minimal visual fidelity loss.

Reviewer: ```suggestion
SPZ is a compression format from Niantic Labs designed for Gaussian Splatting and available as open source under the MIT licensed. SPZ was purpose built for Gaussian Splatting and offers a balance of high compression with minimal visual fidelity loss.
```

We can't mention Scaniverse, and we should avoid terms like "Currently" in any specifications.

This extension allows glTF to support streaming SPZ compressed data instead of uncompressed or meshopt compressed Gaussian splat data.

> ## Adding SPZ compressed data to Primitives

Reviewer: ```suggestion
## Adding Gaussian Splats compressed with SPZ to Primitives
```

> 

Reviewer: Define what a Gaussian Splat is composed of here. Again, we can probably use our old specification to help.

If a primitive contains an `extension` property which defines `KHR_spz_compression` then SPZ compression is required. At this time, there is no requirement for a backup uncompressed buffer.

The extension must be listed in `extensionsUsed` along with `KHR_gaussian_splatting`

```json
  "extensionsUsed" : [

>     "KHR_spz_compression", "KHR_gaussian_splatting"

Reviewer: Is this not meant to be a stand-alone Gaussian Splatting extension?

Right now, I am not sure if having two separate extensions is the correct approach. A single extension that both adds support for Gaussian Splatting and uses the SPZ compression is the current intent that I think we are going for.

      "count": 590392,
      "type": "VEC4"
    }],
```

At minimum accessors will be defined for `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`. Each must have `componentType`, `count`, and `type` defined.

Accessor `type` is defined for the resulting type after decompression and dequantization has occurred.

The accessor `count` must match `numPoints` in `KHR_spz_compression`.

## Conformance

The recommended process for handling SPZ compression is as follows:

- If the loader does not support `KHR_spz_compression`, it must fail.
- If the load does support `KHR_spz_compression` then:
  - The loader must process `KHR_spz_compression` data first. The loader must get the data from `KHR_spz_compression`'s `bufferView` property.
  - The loader then must process `attributes` of the `primitive`. When processing the loader must ignore any `bufferView` and `byteOffset` in the `accessor` and instead use values derived from the decompressed data streams. This data can be used to populate the `accessors` or render directly.
  - Any attributes not listed in `KHR_spz_compression`'s `attributes` but are listed in the `primitive` must then be processed by the loader.

> ## Implementation

Reviewer: We should include all of the relevant stuff from our implementation section of the last proposal, including how to sort the splats and rasterize them.
